dipy/align/imaffine.py

```diff
@@ -0,0 +1,715 @@
+            histograms
+        sampling_proportion : int in (0,100]
+            the percentage of voxels to be used for estimating the (joint and
+            marginal) intensity histograms. If None, dense sampling is used.
+        """
+        super(MattesMIMetric, self).__init__(nbins)
+        self.sampling_proportion = sampling_proportion
+
+    def setup(self, transform, static, moving, static_grid2space=None,
+              moving_grid2space=None, prealign=None):
+        r""" Prepares the metric to compute intensity densities and gradients
+
+        The histograms will be setup to compute probability densities of
+        intensities within the minimum and maximum values of `static` and
+        `moving`
+
+
+        Parameters
+        ----------
+        transform: instance of Transform
+            the transformation with respect to whose parameters the gradient
+            must be computed
+        static : array, shape (S, R, C) or (R, C)
+            static image
+        moving : array, shape (S', R', C') or (R', C')
+            moving image
+        static_grid2space : array (dim+1, dim+1)
+            the grid-to-space transform of the static image
+        moving_grid2space : array (dim+1, dim+1)
+            the grid-to-space transform of the moving image
+        prealign : array, shape (dim+1, dim+1)
+            the pre-aligning matrix (an affine transform) that roughly aligns
+            the moving image towards the static image. If None, no
+            pre-alignment is performed. If a pre-alignment matrix is available,
+            it is recommended to directly provide the transform to the
+            MattesMIMetric instead of manually warping the moving image and
+            provide None or identity as prealign. This way, the metric avoids
+            performing more than one interpolation.
+        """
+        self.dim = len(static.shape)
+        if moving_grid2space is None:
+            moving_grid2space = np.eye(self.dim + 1)
+        if static_grid2space is None:
+            static_grid2space = np.eye(self.dim + 1)
+        self.transform = transform
+        self.static = np.array(static).astype(np.float64)
+        self.moving = np.array(moving).astype(np.float64)
+        self.static_grid2space = static_grid2space
+        self.static_space2grid = npl.inv(static_grid2space)
+        self.moving_grid2space = moving_grid2space
+        self.moving_space2grid = npl.inv(moving_grid2space)
+        self.static_direction, self.static_spacing = \
+            get_direction_and_spacings(static_grid2space, self.dim)
+        self.moving_direction, self.moving_spacing = \
+            get_direction_and_spacings(moving_grid2space, self.dim)
+        self.prealign = prealign
+        self.param_scales = None
+        static_32 = np.array(static).astype(np.float32)
+        moving_32 = np.array(moving).astype(np.float32)
+
+        T = np.eye(self.dim + 1)
+        if self.prealign is not None:
```

Comment:
Once the prealignment happens, do you need to set it to None? Or to `np.eye`? Will it subseqeuntly get reapplied otherwise (e.g. below in `update_dense`)?